File: competitive-programming/src/edu/algos/li02_matrix/medium/L74_SearchA2DMatrix.java
package edu.algos.li02_matrix.medium;

public class L74_SearchA2DMatrix {

	public static void main(String[] args) {
		L74_SearchA2DMatrix obj = new L74_SearchA2DMatrix();
		int[][] matrix = new int[][] { 
            { 1, 4, 7, 11, 15 }, 
            { 2, 5, 8, 12, 19 }, 
            { 3, 6, 9, 16, 22 },
            { 10, 13, 14, 17, 24 }, 
            { 18, 21, 23, 26, 30 } };

		System.out.println(obj.searchMatrix(matrix, 5));
	}

    // Problem: https://leetcode.com/problems/search-a-2d-matrix/
    // Idea: Use binary search from top-right corner
    public boolean searchMatrix(int[][] matrix, int target) {
        int i=0, j=matrix[0].length-1;
        while(j>=0 && i<matrix.length){
            int num = matrix[i][j];
            if(num == target){
                return true;
            }else if(target < num){
                j--;
            }else{
                i++;
            }
        }
        return false;
    }

}

File: dsa/patterns/05_binarysearch/python/l74_search_a2_dmatrix.py
# LeetCode 74 - Search a 2D Matrix
#
# Idea:
# Treat the 2D matrix as a sorted 1D array.
# Apply binary search on indices [0, m*n - 1].
#
# Mapping:
# row = mid // n
# col = mid % n
#
# Time Complexity: O(log(m*n))
# Space Complexity: O(1)

def search_matrix(matrix: list[list[int]], target: int) -> bool:
    rows, cols = len(matrix), len(matrix[0])
    left, right = 0, rows * cols - 1

    while left <= right:
        mid = (left + right) // 2
        row, col = divmod(mid, cols)

        if matrix[row][col] == target:
            return True
        if matrix[row][col] < target:
            left = mid + 1
        else:
            right = mid - 1

    return False


# Driver code
if __name__ == "__main__":
    matrix = [
        [1, 3, 5, 7],
        [10, 11, 16, 20],
        [23, 30, 34, 60]
    ]
    print(search_matrix(matrix, 5))  # True
